lib/rethinkdb/connection.ex

```diff
@@ -123,213 +119,101 @@ defmodule RethinkDB.Connection do
   end
 
-  def handle_call(_, _,
-    state = %{pending: pending, config: %{max_pending: max_pending}}) when map_size(pending) > max_pending do
-    {:reply, %RethinkDB.Exception.TooManyRequests{}, state}
-  end
-
-  def handle_call({:query_noreply, query}, _from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    {:noreply, state} = Request.make_request(query, token, :noreply, %{state | token: new_token})
-    {:reply, :noreply, state}
-  end
-
-  def handle_call({:query, query}, from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    Request.make_request(query, token, from, %{state | token: new_token})
-  end
-
-  def handle_call({:continue, token}, from, state) do
-    query = "[2]"
-    Request.make_request(query, token, from, state)
-  end
+  #
+  # DBConnection API
+  #
 
-  def handle_call({:stop, token}, from, state) do
-    query = "[3]"
-    Request.make_request(query, token, from, state)
+  def connect(options) do
+    # Starts the multiplexer process and stores its pid and some default options to the state.
+    {:ok, pid} = Multiplexer.start_link(options)
+    {:ok, %__MODULE__{pid: pid, options: Keyword.take(options, ~w(db database timeout)a)}}
   end
 
-  def handle_call(:noreply_wait, from, state = %{token: token}) do
-    query = "[4]"
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    Request.make_request(query, token, from, %{state | token: new_token})
+  def disconnect(_err, _state) do
+    # TODO
   end
 
-  def handle_cast(:stop, state) do
-    {:disconnect, :normal, state};
+  def checkin(state) do
+    {:ok, state}
   end
 
-  def handle_info({proto, _port, data}, state = %{socket: socket}) when proto in [:tcp, :ssl] do
-    :ok = Transport.setopts(socket, [active: :once])
-    Request.handle_recv(data, state)
+  def checkout(state) do
+    {:ok, state}
   end
 
-  def handle_info({closed_msg, _port}, state) when closed_msg in [:ssl_closed, :tcp_closed] do
-    {:disconnect, closed_msg, state}
-  end
+  def handle_execute(%Q{message: message} = query, _params, options, %__MODULE__{pid: pid} = state) do
+    # Overrides default options with specified ones.
+    options = Keyword.merge(state.options, options)
 
-  def handle_info(msg, state) do
-    Logger.debug("Received unhandled info: #{inspect(msg)} with state #{inspect state}")
-    {:noreply, state}
-  end
-
-  def terminate(_reason, %{socket: socket}) do
-    Transport.close(socket)
-    :ok
-  end
+    message =
+      unless message do
+        # This only applies when called with run/3 and is used
+        # to support default options when encoding the query.
+        message = describe(query, options)
+        |> Map.fetch!(:message)
+      else
+        message
+      end
 
-  def terminate(_reason, _state) do
-    :ok
+    case Multiplexer.send_recv(pid, message, options) do
```

Comment:
It does not prevent multiplexing because the `Multiplexer` process has it own request/reply loop. Mapping calling process ids with RethinkDB tokens.

So multiple clients can have concurrent queries running at the same time.